pkg/httpconfig/http.go

```diff
@@ -74,106 +67,8 @@ func (b BasicAuth) IsZero() bool {
-			MaxIdleConns:          transportConfig.MaxIdleConns,
-			MaxIdleConnsPerHost:   transportConfig.MaxIdleConnsPerHost,
-			MaxConnsPerHost:       transportConfig.MaxConnsPerHost,
-			TLSClientConfig:       tlsConfig,
-			DisableCompression:    transportConfig.DisableCompression,
-			IdleConnTimeout:       time.Duration(transportConfig.IdleConnTimeout),
-			ResponseHeaderTimeout: time.Duration(transportConfig.ResponseHeaderTimeout),
-			ExpectContinueTimeout: time.Duration(transportConfig.ExpectContinueTimeout),
-			TLSHandshakeTimeout:   time.Duration(transportConfig.TLSHandshakeTimeout),
-			DialContext: conntrack.NewDialContextFunc(
-				conntrack.DialWithTracing(),
-				conntrack.DialWithName(name)),
-		}
-
-		// HTTP/2 support is golang has many problematic cornercases where
-		// dead connections would be kept and used in connection pools.
-		// https://github.com/golang/go/issues/32388
-		// https://github.com/golang/go/issues/39337
-		// https://github.com/golang/go/issues/39750
-		// TODO: Re-Enable HTTP/2 once upstream issue is fixed.
-		// TODO: use ForceAttemptHTTP2 when we move to Go 1.13+.
-		err := http2.ConfigureTransport(rt.(*http.Transport))
-		if err != nil {
-			return nil, err
-		}
-
-		// If a authorization_credentials is provided, create a round tripper that will set the
-		// Authorization header correctly on each request.
-		if cfg.Authorization != nil && len(cfg.Authorization.Credentials) > 0 {
-			rt = config_util.NewAuthorizationCredentialsRoundTripper(cfg.Authorization.Type, cfg.Authorization.Credentials, rt)
-		} else if cfg.Authorization != nil && len(cfg.Authorization.CredentialsFile) > 0 {
-			rt = config_util.NewAuthorizationCredentialsFileRoundTripper(cfg.Authorization.Type, cfg.Authorization.CredentialsFile, rt)
-		}
-		// Backwards compatibility, be nice with importers who would not have
-		// called Validate().
-		if len(cfg.BearerToken) > 0 {
-			rt = config_util.NewAuthorizationCredentialsRoundTripper("Bearer", cfg.BearerToken, rt)
-		} else if len(cfg.BearerTokenFile) > 0 {
-			rt = config_util.NewAuthorizationCredentialsFileRoundTripper("Bearer", cfg.BearerTokenFile, rt)
-		}
-
-		if cfg.BasicAuth != nil {
-			rt = config_util.NewBasicAuthRoundTripper(cfg.BasicAuth.Username, cfg.BasicAuth.Password, cfg.BasicAuth.PasswordFile, rt)
-		}
-		// Return a new configured RoundTripper.
-		return rt, nil
-	}
-
-	tlsConfig, err := config_util.NewTLSConfig(&cfg.TLSConfig)
-	if err != nil {
-		return nil, err
-	}
-
-	if len(cfg.TLSConfig.CAFile) == 0 {
-		// No need for a RoundTripper that reloads the CA file automatically.
-		return newRT(tlsConfig)
-	}
-
-	return config_util.NewTLSRoundTripper(tlsConfig, cfg.TLSConfig.CAFile, newRT)
-}
-
 // NewHTTPClient returns a new HTTP client.
-func NewHTTPClient(cfg ClientConfig, name string) (*http.Client, error) {
+func NewHTTPClient(cfg ClientConfig, name string, metrics *extpromhttp.ClientMetrics) (*http.Client, error) {
```

Comment:
since we also need Clientmetrics now, wanted to understand how to get metrics for sidecar and promclient(error: insufficient args in NewHTTPClient) :slightly_smiling_face: